src/main/java/soot/dexpler/DexFillArrayDataTransformer.java

```diff
@@ -135,4 +113,50 @@ protected void internalTransform(final Body body, String phaseName, Map<String,
     }
   }
 
+  /**
+   * Check the all available definitions of the current array to detect the array type and thus the type of the data loaded
+   * by the array-fill-data instruction.
+   * 
+   * @param l
+   *          local the array we are interested in is saved in
+   * @param u
+   *          unit we start our search
+   * @param defs
+   * @param arrayTypes
+   *          result list containing the discovered array type(s)
+   */
+  private void checkArrayDefinitions(Local l, Unit u, LocalDefs defs, List<Type> arrayTypes) {
+    List<Unit> assDefs = defs.getDefsOfAt(l, u);
+    for (Unit d : assDefs) {
+      if (d instanceof AssignStmt) {
+        AssignStmt arrayAssign = (AssignStmt) d;
+        Value source = arrayAssign.getRightOp();
+        if (source instanceof NewArrayExpr) {
+          // array is assigned from a newly created array
+          NewArrayExpr newArray = (NewArrayExpr) source;
+          arrayTypes.add(newArray.getBaseType());
+          continue;
+        }
+        if (source instanceof InvokeExpr) {
```

Comment:
I'd use an ``else`` here to avoid duplicacte checks.

Reply:
Done